notifications/postgres/notificator.go

```diff
@@ -0,0 +1,336 @@
+		consumers:         make(consumers),
+		storage:           ns,
+		lastKnownRevision: invalidRevisionNumber,
+		revisionMutex:     &sync.RWMutex{},
+	}, nil
+}
+
+func (n *notificator) Start(ctx context.Context, group *sync.WaitGroup) error {
+	if n.ctx != nil {
+		return errors.New("notificator already started")
+	}
+	n.ctx = ctx
+	if err := n.openConnection(); err != nil {
+		return fmt.Errorf("could not open connection to database %v", err)
+	}
+	n.group = group
+	group.Add(1)
+	go n.awaitTermination()
+	return nil
+}
+
+func (n *notificator) RegisterConsumer(userContext *web.UserContext) (notifications.NotificationQueue, int64, error) {
+	queue, err := notifications.NewNotificationQueue(n.queueSize)
+	if err != nil {
+		return nil, invalidRevisionNumber, err
+	}
+	platform := &types.Platform{}
+	err = userContext.Data.Data(platform)
+	if err != nil {
+		return nil, invalidRevisionNumber, fmt.Errorf("could not get platform from user context %v", err)
+	}
+	if platform.ID == "" {
+		return nil, invalidRevisionNumber, errors.New("platform ID not found in user context")
+	}
+	n.isRunningMutex.RLock()
+	defer n.isRunningMutex.RUnlock()
+	if !n.isRunning {
+		return nil, invalidRevisionNumber, errors.New("cannot register consumer - notificator is not running")
+	}
+	if err := n.startListening(); err != nil {
+		return nil, invalidRevisionNumber, fmt.Errorf("listen to %s channel failed %v", postgresChannel, err)
+	}
+
+	n.revisionMutex.RLock()
+	defer n.revisionMutex.RUnlock()
+	n.addConsumer(platform.ID, queue)
+	return queue, n.lastKnownRevision, nil
+}
+
+func (n *notificator) UnregisterConsumer(queue notifications.NotificationQueue) error {
+	n.consumersMutex.Lock()
+	defer n.consumersMutex.Unlock()
+
+	consumerIndex, platformIDToDelete := n.findConsumer(queue.ID())
+	if consumerIndex == -1 {
+		return fmt.Errorf("consumer %s was not found", queue.ID())
+	}
+	platformConsumers := n.consumers[platformIDToDelete]
+	n.consumers[platformIDToDelete] = append(platformConsumers[:consumerIndex], platformConsumers[consumerIndex+1:]...)
+	queue.Close()
+
+	if len(n.consumers[platformIDToDelete]) == 0 {
+		delete(n.consumers, platformIDToDelete)
+	}
+	if len(n.consumers) == 0 {
+		return n.stopListening()
+	}
+	return nil
+}
+
+func (n *notificator) addConsumer(platformID string, queue notifications.NotificationQueue) {
+	n.consumersMutex.Lock()
+	defer n.consumersMutex.Unlock()
+
+	n.consumers[platformID] = append(n.consumers[platformID], queue)
+}
+
+func (n *notificator) findConsumer(id string) (int, string) {
+	var platformIDToDelete string
+	consumerIndex := -1
+	for platformID, platformConsumers := range n.consumers {
+		for index, consumer := range platformConsumers {
+			if consumer.ID() == id {
+				consumerIndex = index
+				break
+			}
+		}
+		if consumerIndex != -1 {
+			platformIDToDelete = platformID
+			break
+		}
+	}
+	return consumerIndex, platformIDToDelete
+}
+
+func (n *notificator) closeAllConsumers() {
```

Comment:
when the Notifier calls closeConsumers it knows the consumers have to be closed and removed. Why do you expect an UnregisterConsumer call  as well ? You reattempt closing each consumer again after "UnregisterConsumer" (which after the isClosed fix won't panic) but is still odd.

Reply:
True. Will be fixed in a different PR.